### PR TITLE
Planner: Depth dependent setpoint changes in CCR mode

### PR DESCRIPTION
**Pull request by atdotde**
We had a user request to allow for setpoint changes
at certain depths for CCR deco.

You can now enter a cylinder with name like
"SP 1.4" ('S' and 'P' and ' ' and a float) with
a switch depth and that cylinder is interpreted as
a depth dependent setpoint switch.

This user interface is a hack. But I believe that such
setpoint changes are similar enough to gas switches during
deco and should thus be handled in a simiar manner.

I would be happy to hear ideas how this could be made
less easter eggish.

Signed-off-by: Robert C. Helling <helling@atdotde.de>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### CODE REVIEW

File: core/planner.c [hunk 1]
```diff
@@ -382,15 +382,31 @@ struct gaschanges {
 	int gasidx;
 };
 
-static struct gaschanges *analyze_gaslist(struct diveplan *diveplan, struct dive *dive, int *gaschangenr, int depth, int *asc_cylinder)
+// Return new setpoint if cylinderi is a setpoint change an 0 if not
+
+static int setpoint_change(struct dive *dive, int cylinderid)
+{
+	cylinder_t *cylinder = get_cylinder(dive, cylinderid);
+	if (cylinder->type.description[0] == 'S' &&
```
Comment (bstoeger) [inline core/planner.c]: Not a C-programmer, but I believe that `strncmp(cylinder->type.description, "SP ", 3) == 0` would be more idiomatic and also not invoke undefined behavior, should the length of `cylinder->type.description` be less than three.

Can `cylinder->type.description` ever be null?

**Comment by atdotde** [comment]
Addressed @bstoeger 's good remarks

**Review by dirkhh — COMMENTED** [review]
I think this makes sense.
The parsing of the cylinder name is indeed a hack (and not all that robust), but given that this is likely for a small audience, I'm fine with it.
What this does need is a CHANGELOG entry and most importantly documentation in the user manual, though.

**Comment by bstoeger** [comment]
The UI is clearly very weird. Just for my understanding: how does that differ from the current "Add set point change" feature?

**Comment by atdotde** [comment]
The difference is that you can add a set point change only for the manually entered part of the dive. The new feature makes the planner add a set point change during the generated ascent.

**Comment by justin-ashworth-research** [comment]
I tested this, it does work. The user also has to realize that the depth-dependence will be set according to the %O2 given for the "SP" "gas."

Much quicker than a C++ refactor in which Diveplan becomes a class that privately manages all access to divedatapoint(s), as well as handling the accounting of duration vs. runtime--which would be one cleaner way to get "timeless" steps into the dive plan logic.

A remaining issue is that the CNS% and OTU are over-calculated for the steps shallower than 6 metres. The logic in the new function needs to automatically enforce that the maximum setpoint/ppO2 at 3 metres is actually 1.3 (100% * 1.3 ata)

**Comment by atdotde** [comment]
Now with documentation (both manual and CHANGELOG). Turns out, two tests were affected by the change to OTU and CNS calculations (the part where ata was changed to bar, I don't know why we ever had ata there).

Fixes #3182

**Comment by justin-ashworth-research** [comment]
That's great, I checked and this additional shallow depth correction is working in the planner (elegantly enough)

A remaining item in those functions calculate_otu() and calculate_cns_dive() is that the if/else and "trueo2" logics miss the correct OC gas po2 when the dive mode is bailout. I managed to get it counted right in these fork commits https://github.com/justin-ashworth-research/subsurface/commit/12837e16f66dbdcd9c2b74761688d649a4b32f6c and https://github.com/justin-ashworth-research/subsurface/commit/6640b01162f9e8c124e3448ee8a64937836d8545 (small fix for the first commit)

**Comment by justin-ashworth-research** [comment]
I didn't even realize that I could edit gas switch depths directly until now. Thanks for the documentation 🙏

**Review by justin-ashworth-research — COMMENTED** [review]
calculate_otu() has a similar issue, I think.

I was checking dc->divemode for each/all of the if() conditions that check setpoints--not just the first case (sample->o2sensor)

trueo2 seems an unnecessary variable, if the compound if/else statement(s) are simplified into a single level.

**Comment by atdotde** [comment]
@dirkhh I think this openSUSE fail is cause by other things and this is ready to merge.

**Review by dirkhh — APPROVED** [review]
This looks good and should be ready to merge

**Pull request merged**